src/main/scala/com/microsoft/ml/spark/explainers/RegressionBase.scala

```diff
@@ -0,0 +1,108 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark.explainers
+
+import breeze.linalg.{*, sum, DenseMatrix => BDM, DenseVector => BDV}
+import breeze.numerics.sqrt
+
+case class RegressionResult(coefficients: BDV[Double], intercept: Double, rSquared: Double, loss: Double)
+  extends (BDV[Double] => Double) {
+  def apply(x: BDV[Double]): Double = coefficients.dot(x) + intercept
+}
+
+/**
+  * The RegressionBase class centers and rescales the input matrix and output vector to support fitting intercept and
+  * specifying sampleWeights. The underlying regression algorithm does not need to support fitting intercept and sample
+  * weights.
+  */
+//noinspection ScalaStyle
+abstract class RegressionBase {
+  def fit(data: BDM[Double], outputs: BDV[Double], fitIntercept: Boolean): RegressionResult = {
+    val weights = BDV.ones[Double](data.rows)
+    fit(data, outputs, weights, fitIntercept)
+  }
+
+  def fit(data: BDM[Double], outputs: BDV[Double], sampleWeights: BDV[Double], fitIntercept: Boolean)
+  : RegressionResult = {
+    require(data.rows == sampleWeights.size, s"data.rows != sampleWeights.size: ${data.rows} != ${sampleWeights.size}")
+    require(sampleWeights.forall(_ >= 0), "Weights must be non-negative")
+
+    val normalizedWeights = normalizeSampleWeights(sampleWeights)
+
+    val (x, y, xOffset, yOffset) = if (fitIntercept) {
+      // step 1: center x and y.
+      val x_weighted = BDM.tabulate(data.rows, data.cols) {
+        (i, j) =>
+          normalizedWeights(i) * data(i, j)
+      }
+
+      val x_offset = (sum(x_weighted(::, *)) /:/ sum(normalizedWeights)).t
```

Comment:
nit: can we try to make these inner vars have a more idiomatic naming scheme?

Reply:
done.